Add support for Phabricator Plugin

laurentgoudet opened the pull request:
I'm trying to add the support for the Phabricator-Jenkins plugin, however I'm not really sure why the build fails:

```
javaposse.jobdsl.dsl.DslScriptLoaderSpec > use @Grab FAILED
    org.codehaus.groovy.control.MultipleCompilationErrorsException at DslScriptLoaderSpec.groovy:183
```

My first guess is that it's due to the plugin not being officially hosted yet, as the registration is in progress (see https://github.com/uber/phabricator-jenkins-plugin/issues/8).

Thoughts?

daspilker commented:
I'll leave this open until the plugin is available in the update center.

sectioneight commented on docs/Job-reference.md:
```diff
@@ -4255,6 +4290,40 @@ The arguments here are in order:
+        phabricatorNotifier {
+            commentOnSuccess(boolean comment = true)                // defaults to false
+            commentWithConsoleLinkOnFailure(boolean comment = true) // defaults to false
+            coberturaReportFile(String coverageFile)                // defaults to false
```
this was removed in a recent refactor (it re-uses the results of the cobertura report, so it doesn't take any configuration)

sectioneight commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PhabricatorNotifierContext.groovy:
```diff
@@ -0,0 +1,28 @@
+        this.commentWithConsoleLinkOnFailure = commentWithConsoleLinkOnFailure
+    }
+
+    void coberturaReportFile(String coberturaReportFile) {
```
unused

sectioneight commented on docs/Job-reference.md:
```diff
@@ -4255,6 +4290,40 @@ The arguments here are in order:
+            commentOnSuccess(boolean comment = true)                // defaults to false
+            commentWithConsoleLinkOnFailure(boolean comment = true) // defaults to false
+            coberturaReportFile(String coverageFile)                // defaults to false
+            enableUberalls(boolean enable = true)                   // defaults to false
```
there is also a new option to drop text into a file (default `.phabricator-comment`) to post back:

https://github.com/uber/phabricator-jenkins-plugin#additional-comments

https://github.com/uber/phabricator-jenkins-plugin/blob/275cc0cf46c65305aba1e7c587967d7a8af59e20/src/main/resources/com/uber/jenkins/phabricator/PhabricatorNotifier/config.jelly#L20-L26

sectioneight commented:
I just pushed 1.0 to maven... still waiting for it to show up on the wiki page https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin ("No Information For This Plugin") but we're getting closer...

laurentgoudet commented:
The plugin is now officially hosted (https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin), and I've updated the PR based on @sectioneight feedback. Tests are still failing though, not sure if I understand how the `@RequiresPlugin` annotation is meant to work :/. Do I need to manually install the plugin with some other means in order to run the tests? Thanks!

laurentgoudet commented:
So it look the `@Grab FAILED` error was due to some corrupted `grape` cache directory on my machine - I squashed & rebased, should be all good now.

daspilker commented on job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContextSpec.groovy:
```diff
@@ -600,6 +600,53 @@ class WrapperContextSpec extends Specification {
+        1 * mockJobManagement.requireMinimumPluginVersion('phabricator-plugin', '1.0')
+    }
+
+    def 'recall preBuildCleanup with minimal options'() {
```
why did you include this test?

laurentgoudet replied:
hmm my bad, bad merge :/

joshuaspence commented:
+1 on support for this plugin

KostyaSha commented:
@daspilker why plugins does not implementing this on their code side?

lanwen commented:
@KostyaSha Now it have [issue JENKINS-28408](https://issues.jenkins-ci.org/browse/JENKINS-28408)

daspilker commented:
There have been 6 releases of the Phabricator plugin in the last 8 days. I will leave this open until the plugin has reached a stable phase. Please check whether the PR needs to updated for the latest release of the Phabricator plugin.

Or move the DSL to the Phabricator plugin, see https://github.com/jenkinsci/job-dsl-plugin/wiki/Extending-the-DSL. JENKINS-28408 will be fixed within the next couple of days.

sectioneight commented:
Yea, we've been using this internally at Uber for about 6 months but it's
by no means hardened for general use yet. Hopefully as more people use it
we can iron out the kinks.

On Wed, Jun 10, 2015 at 8:28 AM Daniel Spilker <notifications@github.com>
wrote:

> There have been 6 releases of the Phabricator plugin in the last 8 days. I
> will leave this open until the plugin has reached a stable phase. Please
> check whether the PR needs to updated for the latest release of the
> Phabricator plugin.
>
> Or move the DSL to the Phabricator plugin, see
> https://github.com/jenkinsci/job-dsl-plugin/wiki/Extending-the-DSL.
> JENKINS-28408 will be fixed within the next couple of days.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/job-dsl-plugin/pull/494#issuecomment-110796134>
> .
>

joshuaspence commented:
Bump. What's the status on this? Is there any interest in having this merged?

daspilker commented:
Please rebase to master, adapt to the latest version of the Phabricator plugin if necessary and adapt to the new documentation style. Otherwise I will close this pull request.

laurentgoudet commented:
Sorry I got sidetracked on that. Give me few days max and I'll fix it.

laurentgoudet commented:
- Rebased on master
- Updated Phabricator plugin to latest version
- Added the new plugin's options
- Updated the unit tests
- Transferred the doc to the new style

daspilker commented on docs/Home.md:
```diff
@@ -54,6 +54,7 @@ Have a look at the [Jenkins Job DSL Gradle example](https://github.com/sheehan/j
  * Added support for the [Progress Bar Column Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Progress+Bar+Column+Plugin)
  * Added support for the [Rebuild Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin)
  * Added support for the [Global Variable String Parameter Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Global+Variable+String+Parameter+Plugin)
+ * Added support for the [Phabricator Plugin](https://github.com/uber/phabricator-jenkins-plugin)
```
link to the Jenkins wiki, not to GitHub

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PhabricatorNotifierContext.groovy:
```diff
@@ -0,0 +1,38 @@
+    int commentSize = 1000
+    boolean commentWithConsoleLinkOnFailure
+
+    void commentOnSuccess(boolean commentOnSuccess = true) {
```
Add GroovyDoc for all DSL methods

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy:
```diff
@@ -1579,6 +1579,26 @@ class PublisherContext extends AbstractExtensibleContext {
+     *
+     * @since 1.39
+     */
+    @RequiresPlugin(id = 'phabricator-plugin', minimumVersion = '1.5')
```
1.5 is not the latest version. Are you sure you want to support 1.5? The options in 1.5 do not match your code.

laurentgoudet replied:
Hmm weird https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin "Latest Release" shows 1.5 even though the last published one is 1.8.1.

laurentgoudet commented:
- Fixed plugin URL
- Added GroovyDoc strings
- Fixed plugin version to v1.8.1

daspilker commented:
Thanks!

@sectioneight: The commitSize option is broken, it can't be set by the DSL or by the web UI. Must be a problem in the Phabricator plugin. It's also very confusing that there are two wiki pages, https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin and https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Differential+Plugin. Can you fix that?

sectioneight commented:
@daspilker just got back from vacation and catching up on things.

What do you mean by "commitSize"? That's not an option provided by the plugin. Comment size? Can you tell me what the observed and expected behavior is?

As far as the wiki goes, that's been bothering me as well. It's an artifact of renaming the plugin for clarity, and triggering a bug in JIRA.

daspilker commented:
@sectioneight ah, sorry, I meant commentSize. The value is always set back to 1000 after saving and reopening the config. You should send an email to the Jenkins dev list to get the wiki page deleted.

sectioneight commented:
Request to delete here: https://groups.google.com/forum/#!topic/jenkinsci-dev/Nd-ErS21USA

I will look into the commentSize issue. I have never use the jobs-dsl-plugin before. Can you give me an example job configuration that reproduces this issue?

daspilker commented:
You don't need the job-dsl-plugin to reproduce the issue. Just create a new job, add the "Post to Phabricator" post-build action, set "Maximum comment character length" to 2000 and click "Save". Then click "Configure" to inspect the config and notice that "Maximum comment character length" is set to 1000.

sectioneight commented:
Whelp, that's embarrassing. Fix in https://github.com/uber/phabricator-jenkins-plugin/pull/105
